### PR TITLE
Change getBaselineAasignment/getBestPossibleAssignment to account for partial-WAGED clusters

### LINKED ISSUE #1166: Change getBaselineAssignment/getBestPossibleAssignment to account for partial-WAGED clusters
WAGED Rebalancer by design only supports one-off migration of resources. As a result, its `getBaselineAssignment()` did not consider the case of a cluster that's partially managed by WAGED. It tries to get baselines for all resources from the cluster; if there's no baseline data, it gets current states for all resources as a fallback. It doesn't care for resources that don't have baseline while other resources have baselines - when that's the case, the resources that don't have baseline will not fallback to their current states; they will simply not have baselines. A similar problem happens to `getBestPossibleAssignment()`. 

The logic is now changed to a per resource level. If a resource has baseline, the baseline will be used; else, the current state will be used. `getBestPossibleAssignment()` will be fixed the same way.

### PR DESCRIPTION
**Pull request by NealSun96**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

Fixes #1166 

### Description

- [x] The following tests are written for this issue:
testPartialBaselineAvailability

- [x] Here are some details about my PR, including screenshots of any UI changes:

WAGED Rebalancer by design only supports one-off migration of resources. As a result, its `getBaselineAssignment()` did not consider the case of a cluster that's partially managed by WAGED. It tries to get baselines for all resources from the cluster; if there's no baseline data, it gets current states for all resources as a fallback. It doesn't care for resources that don't have baseline while other resources have baselines - when that's the case, the resources that don't have baseline will not fallback to their current states; they will simply not have baselines. A similar problem happens to `getBestPossibleAssignment()`. 

The logic is now changed to a per resource level. If a resource has baseline, the baseline will be used; else, the current state will be used. `getBestPossibleAssignment()` will be fixed the same way. 

### Tests

- [x] The following is the result of the "mvn test" command on the appropriate module:
```
[ERROR] Tests run: 1155, Failures: 5, Errors: 0, Skipped: 1, Time elapsed: 4,324.66 s <<< FAILURE! - in TestSuite
[ERROR] testResourceSubset(org.apache.helix.tools.TestClusterStateVerifier)  Time elapsed: 1.025 s  <<< FAILURE!
org.apache.helix.HelixException: Failed to create pause signal
        at org.apache.helix.tools.TestClusterStateVerifier.testResourceSubset(TestClusterStateVerifier.java:115)

[ERROR] afterMethod(org.apache.helix.tools.TestClusterStateVerifier)  Time elapsed: 1.062 s  <<< FAILURE!
java.lang.IllegalStateException: ZkClient already closed!
        at org.apache.helix.tools.TestClusterStateVerifier.afterMethod(TestClusterStateVerifier.java:98)

[ERROR] testCustomizedViewAggregation(org.apache.helix.integration.TestCustomizedViewAggregation)  Time elapsed: 12.133 s  <<< FAILURE!
java.lang.AssertionError: expected:<true> but was:<false>
        at org.apache.helix.integration.TestCustomizedViewAggregation.validateAggregationSnapshot(TestCustomizedViewAggregation.java:238)
        at org.apache.helix.integration.TestCustomizedViewAggregation.testCustomizedViewAggregation(TestCustomizedViewAggregation.java:394)

[ERROR] testStateTransitionTimeOut(org.apache.helix.integration.paticipant.TestStateTransitionTimeoutWithResource)  Time elapsed: 36.061 s  <<< FAILURE!
java.lang.AssertionError: expected:<true> but was:<false>
        at org.apache.helix.integration.paticipant.TestStateTransitionTimeoutWithResource.testStateTransitionTimeOut(TestStateTransitionTimeoutWithResource.java:171)

[ERROR] testPeriodicRefresh(org.apache.helix.integration.spectator.TestRoutingTableProviderPeriodicRefresh)  Time elapsed: 2.015 s  <<< FAILURE!
java.lang.AssertionError: expected:<7> but was:<6>
        at org.apache.helix.integration.spectator.TestRoutingTableProviderPeriodicRefresh.testPeriodicRefresh(TestRoutingTableProviderPeriodicRefresh.java:192)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   TestCustomizedViewAggregation.testCustomizedViewAggregation:394->validateAggregationSnapshot:238 expected:<true> but was:<false>
[ERROR]   TestStateTransitionTimeoutWithResource.testStateTransitionTimeOut:171 expected:<true> but was:<false>
[ERROR]   TestRoutingTableProviderPeriodicRefresh.testPeriodicRefresh:192 expected:<7> but was:<6>
[ERROR]   TestClusterStateVerifier.afterMethod:98 » IllegalState ZkClient already closed...
[ERROR]   TestClusterStateVerifier.testResourceSubset:115 » Helix Failed to create pause...
[INFO] 
[ERROR] Tests run: 1155, Failures: 5, Errors: 0, Skipped: 1
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:12 h
[INFO] Finished at: 2020-07-22T18:49:39-07:00
[INFO] ------------------------------------------------------------------------
```
Rerun
```
[INFO] Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 38.505 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 6, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  43.912 s
[INFO] Finished at: 2020-07-22T19:12:42-07:00
[INFO] ------------------------------------------------------------------------
```

### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- [x] My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java [hunk 1]
```diff
@@ -643,23 +643,28 @@ private void validateInput(ResourceControllerDataProvider clusterData,
...
       } catch (Exception ex) {
         throw new HelixRebalanceException(
             "Failed to get the current baseline assignment because of unexpected error.",
             HelixRebalanceException.Type.INVALID_REBALANCER_STATUS, ex);
       }
     }
-    if (currentBaseline.isEmpty()) {
+    currentBaseline.keySet().retainAll(resources);
+    if (!currentBaseline.keySet().containsAll(resources)) {
       LOG.warn("The current baseline assignment record is empty. Use the current states instead.");
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java:660]: The current baseline assignment record does not contain all resources assignment....

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java [hunk 2]
```diff
@@ -643,23 +643,28 @@ private void validateInput(ResourceControllerDataProvider clusterData,
...
         _stateReadLatency.endMeasuringLatency();
       } catch (Exception ex) {
         throw new HelixRebalanceException(
             "Failed to get the current baseline assignment because of unexpected error.",
             HelixRebalanceException.Type.INVALID_REBALANCER_STATUS, ex);
       }
     }
-    if (currentBaseline.isEmpty()) {
+    currentBaseline.keySet().retainAll(resources);
+    if (!currentBaseline.keySet().containsAll(resources)) {
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java:659]: containsAll() will iterate through resources anyway. So just do the loop without check here.

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java [hunk 3]
```diff
@@ -643,23 +643,28 @@ private void validateInput(ResourceControllerDataProvider clusterData,
...
       LOG.warn("The current baseline assignment record is empty. Use the current states instead.");
-      currentBaseline = currentStateOutput.getAssignment(resources);
+      for (Map.Entry<String, ResourceAssignment> entry : currentStateOutput.getAssignment(resources)
+          .entrySet()) {
+        if (resources.contains(entry.getKey()) && !currentBaseline.containsKey(entry.getKey())) {
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java:663]: currentBaseline.putIfAbsent() ?

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java [hunk 4]
```diff
@@ -643,23 +643,28 @@ private void validateInput(ResourceControllerDataProvider clusterData,
...
             "Failed to get the current baseline assignment because of unexpected error.",
             HelixRebalanceException.Type.INVALID_REBALANCER_STATUS, ex);
       }
     }
-    if (currentBaseline.isEmpty()) {
+    currentBaseline.keySet().retainAll(resources);
+    if (!currentBaseline.keySet().containsAll(resources)) {
       LOG.warn("The current baseline assignment record is empty. Use the current states instead.");
-      currentBaseline = currentStateOutput.getAssignment(resources);
+      for (Map.Entry<String, ResourceAssignment> entry : currentStateOutput.getAssignment(resources)
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java:661]: getAssignment() is very expensive, let's do it for the missing resources only. If you do it for all resources, the algorithm performance would be impacted.

Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java:661]: Better to do:
1. find all the missing resources
2. call getAssignment() to get the corresponding assignment
3. add assignments to the return map.

**Comment by NealSun96** [comment]
This PR is ready to be merged, approved by @jiajunwang   
Final commit message:
## Change getBaselineAasignment/getBestPossibleAssignment to account for partial-WAGED clusters ##
The logics are now changed to a per resource level. If a resource has baseline/best possible, the baseline/best possible will be used; else, the current state will be used.

**Pull request merged**